### ISSUE TITLE
table locking on Lustre filesystem

### ISSUE THREAD

**Issue by mknapp55** [casacore/casacore#605]
I ran into an issue with casacore that I suspect is related to the Lustre filesystem on the cluster I'm using. The error I get when running NDPPP is:
```
terminate called after throwing an instance of 'casa::TableError'
  what():  Error (Function not implemented) when acquiring lock on /data/working_L531971/pipeline/L531971_SB470_uv.calibrator-flagcal/table.lock
```
1. Is there a known fix for this?  I found some references to MultiFile when I was googling this problem, but it wasn't clear to me what to do.
2. I'm using Kern to install/update casacore and other packages built on top of casacore.  I first raised this issue with Kern (https://github.com/kernsuite/packaging/issues/50#issuecomment-309682475), and @gijzelaerr suggested I ask if a solution to the Lustre locking issue (if it exists) can be backported to casacore 2.2.x?

**Comment by mpokorny** [comment]
Using CASA with files on a Lustre filesystems is common, so I'd suspect the filesystem or something unusual in what you were doing. Have you checked that the Lustre filesystem is mounted with the "flock" option, or perhaps the "localflock" option? What were you trying when the error occurred?

**Comment by mknapp55** [comment]
I was using NDPPP as part of a pipeline run.  I think this step was just copying data to a new MS.  I have no idea what 'flock' or 'localflock' are, but I can ask my sysadmin tomorrow.  This is a large, multi-tenant HPC, so I have no direct control over the filesystem setup.  I'm confident there's nothing wrong with the code because I was able to run it successfully after I copied my data to a non-Lustre temporary storage area.

LOFAR's wiki recommends turning off file locking for Lustre file systems - see https://www.astron.nl/lofarwiki/doku.php?id=public:install_lofar_centos6.5, section 3.2.  That wiki page is what gave me the hint that Lustre was the problem and I should try moving my data to a non-Lustre partition.

**Comment by mpokorny** [comment]
Lustre filesystems only support POSIX file locking with either the flock or localflock mount options in effect. Since casacore does file locking with fcntl, you might want to check for those mount options. I'm unfamiliar with NDPPP, but might it be possible to use a casacore::TableLock with the NoLocking option? I just noticed in section 3.2 of the LOFAR wiki page you referenced that it's recommended to build casacore with the -DENABLE_TABLELOCKING=NO option for Lustre filesystems. Would that be an option for you?

**Comment by mknapp55** [comment]
I've installed casacore and everything on top of it via [Kern](http://kernsuite.info/), which packages commonly used radio astronomy software so it can be installed via apt-get.  It is possible, I suppose, for me to build casacore from source, but I am very reluctant to do so since so many other tools are built on top of it and I'm almost certain to break something.  Before Kern, I spent weeks attempting to install the LOFAR software stack from source - it never worked.  That's why I originally posted this issue with Kern - I was hoping there was some way I could install their packaged casacore with the `-DENABLE_TABLELOCKING=NO` option (which is obviously non-standard).  I was hoping that there might be some fix in a more recent version of casacore that would allow me to avoid a custom build.

I will ask about the mounting options for my cluster, but I highly doubt the sysadmins will be willing to make any changes since the cluster actively serves many users with a wide variety of software needs.

**Comment by alxames** [comment]
FYI, the LOFAR CEP4 cluster with a Lustre filesys is mounted with the flock option.

Wrt the actual casacore table locking issue, maybe @gervandiepen can shed light on it. Perhaps casacore needs a fix, or maybe NDPPP needs to properly deal with the exception.

**Comment by gervandiepen** [comment]
Casacore has the option to use no table locking. This can be done at
compile time or the table can be opened without locking.

**Comment by gijzelaerr** [comment]
@mknapp55 did @gervandiepen his answer solve your issue?

**Comment by adrianjhpc** [comment]
If I'm only ever reading a table do I need any table locking enabled?

**Comment by gijzelaerr** [comment]
i guess not.

**Comment by gervandiepen** [comment]
No, only when simultaneously reading and writing a table in different
processes you will need locking.
Note that you can build Casacore with table locking disabled.

**Comment by adrianjhpc** [comment]
Thanks both. I've a user case where the read fails (or at least returns zero data) when I turn locking off. I'm just isolating the error at the moment but it looks like it's something to do with the MemoryIO stuff.

**Comment by adrianjhpc** [comment]
I created a fix for it. I just haven't written an isolated test case yet. Will do shortly and submit it with the fix,
